x-pack/legacy/plugins/lens/public/xy_visualization/to_expression.ts

```diff
@@ -122,82 +97,92 @@ export const buildExpression = (
   metadata: Record<string, Record<string, OperationMetadata | null>>,
   frame?: FramePublicAPI,
   { xTitle, yTitle }: { xTitle: string; yTitle: string } = { xTitle: '', yTitle: '' }
-): Ast => ({
-  type: 'expression',
-  chain: [
-    {
-      type: 'function',
-      function: 'lens_xy_chart',
-      arguments: {
-        xTitle: [xTitle],
-        yTitle: [yTitle],
-        legend: [
-          {
-            type: 'expression',
-            chain: [
-              {
-                type: 'function',
-                function: 'lens_xy_legendConfig',
-                arguments: {
-                  isVisible: [state.legend.isVisible],
-                  position: [state.legend.position],
+): Ast | null => {
+  const validLayers = state.layers.filter(layer => layer.xAccessor && layer.accessors.length);
```

Comment:
If we make this function an actual type guard, we can get rid of a lot of non-null assertion operators later down (and we want this, since every non-null assertion operator is potentially a bug later, since we disable the type system at that point).

```ts
interface ValidLayer extends LayerConfig {
  xAccessor: NonNullable<LayerConfig['xAccessor']>;
}
// ...
const validLayers = state.layers.filter((layer): layer is ValidLayer =>
  Boolean(layer.xAccessor && layer.accessors.length)
);
```

Then we can remove all non null assertion operators on `xAccessor` below.

Reply:
Done